fendermint/ipc/src/cache.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: Apache-2.0, MIT
+
+use num_traits::{PrimInt};
+use std::cmp::{max, min};
+use std::collections::HashMap;
+use std::hash::Hash;
+
+type Bounds<T> = (T, T);
+
+/// The key value cache. The key must be numeric and falls within a range/bound.
+pub struct RangeKeyCache<Key, Value> {
+    /// Stores the data in a hashmap.
+    data: HashMap<Key, Value>,
+    /// The lower and upper bound of keys stored in data
+    bounds: Option<Bounds<Key>>,
+}
+
+impl<Key: PrimInt + Hash, Value> RangeKeyCache<Key, Value> {
+    pub fn new() -> Self {
+        Self {
+            data: Default::default(),
+            bounds: None,
+        }
+    }
+
+    pub fn upper_bound(&self) -> Option<Key> {
```

Comment:
I saw later that if you started with an empty cache, inserted something, then removed it, this would still remember what the upper bound used to be. If this is an intended behaviour, it would be great to capture it as an invariant, although personally I'm not sure how intuitive it is that the cache knows about a height which you cannot retrieve from it.